### PR TITLE
chore: cache the total number of lines in the csv file

### PR DESCRIPTION
**Pull request by 0xff-dev**
#### What this PR does / why we need it
chore: cache the total number of lines in the csv file
#### Which issue(s) this PR fixes

Fixes #

### CODE REVIEW

**Review by bjwswang — COMMENTED** [review]
@0xff-dev Please add more comments on your code so people can understand what your are trying to do easily.

**Comment by 0xff-dev** [comment]
> @0xff-dev Please add more comments on your code so people can understand what your are trying to do easily.

Done.

File: apiserver/pkg/common/read_csv.go [hunk 1]
```diff
@@ -20,37 +20,54 @@ import (
...
+type ReadCSVResult struct {
+	Rows  [][]string `json:"rows"`
+	Total int64      `json:"total"`
+}
+
+// ReadCSV Reads the contents of a csv file by lines according to startLine and lines,
+// if cacheLines=0 then the total number of lines in the file will be counted,
+// otherwise the total number of lines in the file will not be counted and the speedup function returns.
+func ReadCSV(o io.Reader, startLine, lines, cacheLines int64) (ReadCSVResult, error) {
+	result := ReadCSVResult{}
```
Comment (bjwswang) [inline apiserver/pkg/common/read_csv.go:32]:
```golang
result := ReadCSVResult{
    Total:  cachedTotalLines
}
```

File: apiserver/pkg/common/read_csv.go [hunk 2]
```diff
@@ -20,37 +20,54 @@ import (
...
-func ReadCSV(o io.Reader, startLine, lines int64) ([][]string, int64, error) {
+type ReadCSVResult struct {
+	Rows  [][]string `json:"rows"`
+	Total int64      `json:"total"`
+}
+
+// ReadCSV Reads the contents of a csv file by lines according to startLine and lines,
+// if cacheLines=0 then the total number of lines in the file will be counted,
+// otherwise the total number of lines in the file will not be counted and the speedup function returns.
+func ReadCSV(o io.Reader, startLine, lines, cacheLines int64) (ReadCSVResult, error) {
```
Comment (bjwswang) [inline apiserver/pkg/common/read_csv.go:31]: `cacheLines` ->  `cachedTotalLines`

File: apiserver/pkg/common/read_csv.go [hunk 3]
```diff
@@ -20,37 +20,54 @@ import (
...
-			result = append(result, line)
-			recordLines++
+			continue
 		}
+
+		result.Rows = append(result.Rows, line)
 	}
 
-	return result, totalLines, err
+	if cacheLines != 0 {
```
Comment (bjwswang) [inline apiserver/pkg/common/read_csv.go:69]: Not required. we can set this `cache` to total at the begining

File: apiserver/pkg/common/read_csv.go (hunk 1 above)
Comment (0xff-dev) [reply apiserver/pkg/common/read_csv.go:32]: https://github.com/kubeagi/arcadia/pull/385/files#diff-5d922defc7220baaf50dc44e878589147fea11600bfacbb79f30301654560991R50

`Total` is used to count the number of lines from the beginning, if the beginning is set to cachedTotallines, cachedTotalLines=0 is normal, but if cachedTotalLines is not 0, the total number of lines will be wrong

File: apiserver/service/minio_server.go [hunk 4]
```diff
@@ -583,6 +582,23 @@ func (m *minioAPI) ReadCSVLines(ctx *gin.Context) {
...
+			"message": err.Error(),
+		})
+	}
+	statInfo := anyStatInfo.(minio.ObjectInfo)
+
+	// checks if the total number of lines in the file has been cached
+	key := [4]string{cachePrefix, bucket, bucketPath, statInfo.ETag}
+	if a, ok := m.lru.Get(key); ok {
+		if v, ok1 := a.(int64); ok1 {
+			klog.Infof("nice, key: %v match, the file has not changed, and the total number of lines in the file is %d", key, v)
```
Comment (bjwswang) [inline apiserver/service/minio_server.go:598]: Use klog.V(5).Info

**Review by bjwswang — APPROVED** [review]

**Review by nkwangleiGIT — APPROVED** [review]

**Pull request merged**